### PR TITLE
BREAKING: React 18 Server Components Compat

### PR DESCRIPTION
**Pull request by LekoArts**
- Don't pass `navigate` through context (as it's a function)
- Convert everything to function components
- Convert everything to hooks
- Optionally use `createServerContext` for context
- Build files with `microbundle`
- Use `@testing-library/react` for tests

This is a **BREAKING** change so we'll need to publish a new major after this.

[ch53629]

### CODE REVIEW

**Comment by tujoworker** [comment]
By now, from code I see that the router when used by projects/app code, gets replaced by a Webpack alias, and the internal router is used. 
Will you consider change that to be a core part in future (breaking)? Like import from “gatsby/router”

**Review by wardpeet — APPROVED** [review]

File: src/lib/location-provider.js [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+            setContext({ location })
+          })
+        })
+      })
+      setRefs({ unlisten })
+
+      return () => {
+        refs.unlisten()
+      }
+    }, [])
```
Comment (wardpeet) [inline src/lib/location-provider.js]: No need to setState for unlisten as it's only used in scope.  I've also added an isCancelled to make sure React don't throw on unmount because setContext is async.

```suggestion
    React.useEffect(() => {
      let isCancelled = false;
      const unlisten = history.listen(() => {
        Promise.resolve().then(() => {
          requestAnimationFrame(() => {
            if (!isCancelled) {
              setContext({ location })
            }
          })
        })
      })
      return () => {
        isCancelled = true;
        unlisten()
      }
    }, [])
```

File: src/lib/location-provider.js [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+export const LocationProvider = withErrorBoundary(
+  ({ history = globalHistory, children }) => {
+    const { location } = history
+
+    const [context, setContext] = React.useState({ location })
+    const [refs, setRefs] = React.useState({ unlisten: null })
+    const [error, _] = useErrorBoundary()
+
+    React.useEffect(() => {
+      history._onTransitionComplete()
```
Comment (wardpeet) [inline src/lib/location-provider.js:15]: This get's called on initial render as well.

File: src/lib/location-provider.js [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+      setRefs({ unlisten })
+
+      return () => {
+        refs.unlisten()
+      }
+    }, [])
+
+    if (error) {
+      if (isRedirect(error)) {
+        navigate(error.uri, { replace: true })
```
Comment (wardpeet) [inline src/lib/location-provider.js:37]: I don't think this is the right pattern, you want to do it inside the errorBoundary or with the onError function

File: src/lib/redirect.js [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+function RedirectImpl(props) {
+  const { to, replace = true, state, noThrow, baseuri } = props
+
+  React.useEffect(() => {
+    Promise.resolve().then(() => {
+      const resolvedTo = resolve(to, baseuri)
+      navigate(insertParams(resolvedTo, props), { replace, state })
+    })
+  }, [resolve, navigate, insertParams])
```
Comment (wardpeet) [inline src/lib/redirect.js]: In the current implementation this is an onMount hook so `[]`.

File: src/lib/focus-handler.js [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+    setFocusHandlerCount(f => f + 1)
+    focus()
+
+    return () => {
+      setFocusHandlerCount(f => f - 1)
+      if (focusHandlerCount === 0) {
+        initialRender.current = true
+      }
+    }
+  }, [])
```
Comment (wardpeet) [inline src/lib/focus-handler.js]: You need to keep focusHandlerCount outside of React as this is being tracked globally
```suggestion
  React.useEffect(() => {
    focusHandlerCount++;
    focus()

    return () => {
      focusHandlerCount--;
      if (focusHandlerCount ===  0) {
        initialRender.current = true
      }
    }
  }, [])
```

File: src/lib/focus-handler.js [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+      if (initialRender.current) {
+        initialRender.current = false
+      } else if (node) {
+        // React polyfills [autofocus] and it fires earlier than cDM, so we were stealing focus away, this line prevents that.
+        if (!node.current.contains(document.activeElement)) {
+          node.current.focus()
+        }
+      }
+    }
+  }
```
Comment (wardpeet) [inline src/lib/focus-handler.js]: Add useCallback to cache the focus handler
```suggestion
  const focus = useCallback(() => {
    if (process.env.NODE_ENV === "test") {
      // TODO: Still a problem?
      // getting cannot read property focus of null in the tests
      // and that bit of global `initialRender` state causes problems
      return
    }

    if (requestFocus) {
      requestFocus(node.current)
    } else {
      if (initialRender.current) {
        initialRender.current = false
      } else if (node) {
        // React polyfills [autofocus] and it fires earlier than cDM, so we were stealing focus away, this line prevents that.
        if (!node.current.contains(document.activeElement)) {
          node.current.focus()
        }
      }
    }
  }, [])
```

File: src/lib/focus-handler.js [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+  uri,
+  location,
+  ...domProps
+}) => {
+  let node = React.useRef()
+
+  const [shouldFocus, setShouldFocus] = React.useState(true)
+  const [prevUri, setPrevUri] = React.useState(uri)
+  const [prevPathname, setPrevPathname] = React.useState(location.pathname)
+  const [focusHandlerCount, setFocusHandlerCount] = React.useState(0)
```
Comment (wardpeet) [inline src/lib/focus-handler.js]:
```suggestion
```

File: src/lib/focus-handler.js [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+    if (!shouldFocus && node) {
+      node.current.focus()
+    }
+  }
+
+  return (
+    <Comp
+      style={{ outline: "none", ...style }}
+      tabIndex="-1"
+      ref={n => (node = n)}
```
Comment (wardpeet) [inline src/lib/focus-handler.js:104]: Correct use of ref
```suggestion
      ref={node}
```

File: src/lib/focus-handler.js [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+  if (uri) {
+    const myURIChanged = uri !== prevUri
+    const navigatedUpToMe =
+      prevPathname !== location.pathname && location.pathname === uri
+
+    setPrevUri(uri)
+    setPrevPathname(location.pathname)
+
+    setShouldFocus(myURIChanged || navigatedUpToMe)
+  }
```
Comment (wardpeet) [inline src/lib/focus-handler.js]: You probably want to do this logic in a useEffect and skip setState.

**Review by wardpeet — COMMENTED** [review]
Added some comments on behavior changes

File: src/lib/focus-handler.js (hunk 8 above)
Comment (LekoArts) [reply src/lib/focus-handler.js:104]: I think this was done on purpose

Comment (LekoArts) [reply src/lib/focus-handler.js:104]: https://blog.thoughtspile.tech/2021/05/17/everything-about-react-refs/

> There’s a subtle difference between having an arrow as your ref prop and a ref object or a stable callback — the arrow has a new identity on every render, forcing the ref to go through an update cycle null. This is normally not too bad, but good to know.

See ref arrows

```jsx
// this does node -> null -> node on every render
<div ref={e => this.node = e} />
// this doesn't
<div ref={useCallback(e => this.node = e, [])} />
// neither does this
setRef = e => this.node = e;
<div ref={this.setRef} />
// this is fine, too
const ref = useRef();
<div ref={ref} />
```

This was added in https://github.com/reach/router/commit/738b3caa41fca4d327aeac7123beeca56ec18554 but no explanation.

Like I knew how to correctly do this, I just wanted to keep old behavior. If we thoroughly test this with all browser, screen readers, keyboard etc. and it works with the above, too, then we can change.

React docs: https://reactjs.org/docs/refs-and-the-dom.html#caveats-with-callback-refs

Comment (wardpeet) [reply src/lib/focus-handler.js:104]: I don't really understand why you want do have that behavior 😅 Cause refs are not tracked by React, there is no real benefit of putting them to null but let's keep it how it is

File: src/lib/location-provider.js (hunk 3 above)
Comment (LekoArts) [reply src/lib/location-provider.js:37]: I'm not sure I'm following. What do you suggest we do here instead? The code I copied has a callback function in the `useErrorBoundary` hook itself (https://github.com/tatethurston/react-use-error-boundary). Do you mean that or something else?

File: package.json [hunk 10]
```diff
@@ -1,91 +1,76 @@
 {
   "name": "@gatsbyjs/reach-router",
-  "version": "1.3.9",
-  "description": "Next generation Routing for React.",
-  "main": "index.js",
-  "module": "es/index.js",
+  "version": "1.3.10-no-navigate.5",
```
Comment (pieh) [inline package.json]: We probably should not change version in this PR - the versioning should be separate commit (and it would "premajor" bump

```suggestion
  "version": "1.3.9",
```

**Review by pieh — APPROVED** [review]
